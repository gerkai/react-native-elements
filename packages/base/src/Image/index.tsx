import { Image as ImageNative } from 'react-native';
import { Image, ImageProps } from './Image';

const DefaultImage = Object.assign(Image, {
  getSize: ImageNative.getSize,
  getSizeWithHeaders: ImageNative.getSizeWithHeaders,
  prefetch: ImageNative.prefetch,
  abortPrefetch: ImageNative.abortPrefetch,
  queryCache: ImageNative.queryCache,
  resolveAssetSource: ImageNative.resolveAssetSource,
});

export { DefaultImage as Image };
export type { ImageProps };
