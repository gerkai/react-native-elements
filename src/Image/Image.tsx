import React, { useCallback } from 'react';
import {
  Animated,
  Image as ImageNative,
  ImageLoadEventData,
  ImageProps as RNImageProps,
  NativeSyntheticEvent,
  StyleProp,
  StyleSheet,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';
import { Text } from '..';
import { RneFunctionComponent } from '../helpers';

export type ImageProps = RNImageProps & {
  Component?: typeof React.Component;
  onPress?(): void;
  onLongPress?(): void;
  ImageComponent?: typeof React.Component;
  PlaceholderContent?: React.ReactElement<any>;
  containerStyle?: StyleProp<ViewStyle>;
  childrenContainerStyle?: StyleProp<ViewStyle>;
  placeholderStyle?: StyleProp<ViewStyle>;
  transition?: boolean;
  transitionDuration?: number;
};

export const Image: RneFunctionComponent<ImageProps> = ({
  onPress,
  onLongPress,
  Component = onPress || onLongPress ? TouchableOpacity : View,
  placeholderStyle,
  PlaceholderContent,
  containerStyle,
  childrenContainerStyle = null,
  style = {},
  ImageComponent = ImageNative,
  onLoad,
  children,
  transition,
  transitionDuration = 360,
  ...props
}) => {
  const root = React.useRef<ImageNative>(null);
  const placeholderOpacity = React.useRef(new Animated.Value(1));

  const onLoadHandler = useCallback(
    (event: NativeSyntheticEvent<ImageLoadEventData>) => {
      if (transition) {
        Animated.timing(placeholderOpacity.current, {
          toValue: 0,
          duration: transitionDuration,
          useNativeDriver: true,
        }).start();
      } else {
        placeholderOpacity.current.setValue(0);
      }
      onLoad?.(event);
    },
    [transition, transitionDuration, onLoad]
  );

  const hasImage = Boolean(props.source);

  return (
    <Component
      onPress={onPress}
      onLongPress={onLongPress}
      accessibilityIgnoresInvertColors={true}
      style={StyleSheet.flatten([styles.container, containerStyle])}
    >
      <ImageComponent
        ref={root}
        testID="RNE__Image"
        {...props}
        {...{ transition, transitionDuration }}
        onLoad={onLoadHandler}
        style={StyleSheet.flatten([StyleSheet.absoluteFill, style])}
      />
      {/* Transition placeholder */}
      <Animated.View
        pointerEvents={hasImage ? 'none' : 'auto'}
        accessibilityElementsHidden={hasImage}
        importantForAccessibility={hasImage ? 'no-hide-descendants' : 'yes'}
        style={[
          StyleSheet.absoluteFillObject,
          {
            opacity: hasImage ? placeholderOpacity.current : 1,
          },
        ]}
      >
        <View
          testID="RNE__Image__placeholder"
          style={StyleSheet.flatten([
            style,
            styles.placeholder,
            placeholderStyle,
          ])}
        >
          {React.isValidElement(PlaceholderContent)
            ? PlaceholderContent
            : PlaceholderContent && (
                <Text testID="RNE__Image__Placeholder__Content">
                  {PlaceholderContent}
                </Text>
              )}
        </View>
      </Animated.View>
      {/* Children for Image */}
      <View
        testID="RNE__Image__children__container"
        style={childrenContainerStyle ?? style}
      >
        {children}
      </View>
    </Component>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'transparent',
    position: 'relative',
    overflow: 'hidden',
  },
  placeholder: {
    backgroundColor: '#bdbdbd',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

Image.displayName = 'Image';
