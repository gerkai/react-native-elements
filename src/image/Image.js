import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {
  Animated,
  Image as ImageNative,
  StyleSheet,
  View,
  Platform,
} from 'react-native';

import { nodeType } from '../helpers';
import { ViewPropTypes, withTheme } from '../config';

const Image = ({
  placeholderStyle,
  PlaceholderContent,
  containerStyle,
  style,
  ImageComponent,
  children,
  ...attributes
}) => {
  const [placeholderOpacity] = useState(new Animated.Value(1));
  const hasImage = typeof attributes.source !== 'undefined';

  const onLoad = () => {
    const minimumWait = 100;
    const staggerNonce = 200 * Math.random();

    setTimeout(
      () => {
        Animated.timing(placeholderOpacity, {
          toValue: 0,
          duration: 350,
          useNativeDriver: Platform.OS === 'android' ? false : true,
        }).start();
      },
      Platform.OS === 'android' ? 0 : Math.floor(minimumWait + staggerNonce)
    );
  };

  return (
    <View
      accessibilityIgnoresInvertColors={true}
      style={StyleSheet.flatten([styles.container, containerStyle])}
    >
      <ImageComponent
        {...attributes}
        onLoad={onLoad}
        style={[
          StyleSheet.absoluteFill,
          {
            width: style.width,
            height: style.height,
          },
        ]}
        testID="RNE__Image"
      />

      <Animated.View
        pointerEvents={hasImage ? 'none' : 'auto'}
        accessibilityElementsHidden={hasImage}
        importantForAccessibility={hasImage ? 'no-hide-descendants' : 'yes'}
        style={[
          styles.placeholderContainer,
          {
            opacity: hasImage ? placeholderOpacity : 1,
          },
        ]}
      >
        <View
          testID="RNE__Image__placeholder"
          style={StyleSheet.flatten([
            style,
            styles.placeholder,
            placeholderStyle,
          ])}
        >
          {PlaceholderContent}
        </View>
      </Animated.View>

      <View style={style}>{children}</View>
    </View>
  );
};

const styles = {
  container: {
    backgroundColor: 'transparent',
    position: 'relative',
  },
  placeholderContainer: {
    ...StyleSheet.absoluteFillObject,
  },
  placeholder: {
    backgroundColor: '#bdbdbd',
    alignItems: 'center',
    justifyContent: 'center',
  },
};

Image.propTypes = {
  ...ImageNative.propTypes,
  ImageComponent: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  PlaceholderContent: nodeType,
  containerStyle: ViewPropTypes.style,
  placeholderStyle: ImageNative.propTypes.style,
};

Image.defaultProps = {
  ImageComponent: ImageNative,
  style: {},
};

export { Image };
export default withTheme(Image, 'Image');
