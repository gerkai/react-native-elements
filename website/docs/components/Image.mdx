---
id: image

title: Image
---

import Usage from "./usage/Image.mdx";

import { Image } from "react-native-elements";

import Tabs from "@theme/Tabs";

import TabItem from "@theme/TabItem";

Drop-in replacement for the standard React Native Image component that displays

images with a placeholder and smooth image load transitioning.

## Usage

<Usage />

## Props

### Image

- [Component](#component)
- [ImageComponent](#imagecomponent)
- [PlaceholderContent](#placeholdercontent)
- [childrenContainerStyle](#childrencontainerstyle)
- [containerStyle](#containerstyle)
- [onLongPress](#onlongpress)
- [onPress](#onpress)
- [onPressIn](#onpressin)
- [onPressOut](#onpressout)
- [placeholderStyle](#placeholderstyle)
- [pressableProps](#pressableprops)
- [transition](#transition)
- [transitionDuration](#transitionduration)

## Reference

### Image

#### Component

Define the component passed to image.

| Type            | Default                                           |
| --------------- | ------------------------------------------------- |
| React Component | `Press handlers present then Pressable else View` |

---

#### ImageComponent

Specify a different component as the Image component.

| Type | Default |
| ---- | ------- |
| any  | None    |

---

#### PlaceholderContent

Content to load when Image is rendering.

| Type               | Default |
| ------------------ | ------- |
| React.ReactElement | None    |

---

#### childrenContainerStyle

Additional styling for the children container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | null    |

---

#### containerStyle

Additional styling for the container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### onLongPress

Called when a long-tap gesture is detected.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### onPress

Called when a single tap gesture is detected.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### onPressIn

Called when a touch is engaged before `onPress`.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### onPressOut

Called when a touch is released before `onPress`.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### placeholderStyle

Additional styling for the placeholder container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### pressableProps

| Type                                 | Default |
| ------------------------------------ | ------- |
| PressableProps except click handlers | None    |

---

#### transition

Perform fade transition on image load.

| Type    | Default |
| ------- | ------- |
| boolean | None    |

---

#### transitionDuration

Perform fade transition on image load.

| Type   | Default |
| ------ | ------- |
| number | 360     |

---
