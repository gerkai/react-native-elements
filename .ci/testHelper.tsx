import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider, FullTheme } from '../src/config';

const theme = {
  colors: {
    black: 'black',
  },
};

export const renderWithTheme = (
  children: any,
  themeProp: Partial<FullTheme> = theme
) => {
  return render(<ThemeProvider theme={themeProp}>{children}</ThemeProvider>);
};
