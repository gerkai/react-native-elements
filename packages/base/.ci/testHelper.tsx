import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import { ThemeProvider, FullTheme, colors } from '../src/config';
import deepmerge from 'deepmerge';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<FullTheme> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    themeProp ? (
      <ThemeProvider theme={deepmerge({ colors }, themeProp)}>
        {children}
      </ThemeProvider>
    ) : (
      children
    ),
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper');
  return { wrapper, ...renderApi };
};
