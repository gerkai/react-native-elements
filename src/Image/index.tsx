import { withTheme } from '../config';
import { Image, ImageProps } from './Image';

export { Image };
export type { ImageProps };
export default withTheme(Image, 'Image');
