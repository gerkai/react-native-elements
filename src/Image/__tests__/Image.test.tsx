import React from 'react';
import { Image as RNImage, Text } from 'react-native';
import Image from '../index';
import { renderWithTheme } from '../../../.ci/testHelper';

jest.useFakeTimers();

describe('Image Component', () => {
  const FAKE_URI = 'https://i.imgur.com/0y8Ftya.jpg';
  it('should render', () => {
    const { toJSON } = renderWithTheme(
      <Image testID="customTestId" source={{ uri: FAKE_URI }} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render the appropriate testId when one is passed.', () => {
    const { queryByTestId } = renderWithTheme(
      <Image testID="customTestId" source={{ uri: FAKE_URI }} />
    );
    const image = queryByTestId('customTestId');
    expect(image.props.source.uri).toBe(FAKE_URI);
  });

  it('should render the appropriate testId when one is not passed.', () => {
    const { queryByTestId } = renderWithTheme(
      <Image source={{ uri: FAKE_URI }} />
    );
    const image = queryByTestId('RNE__Image');
    expect(image.props.source.uri).toBe(FAKE_URI);
  });

  it('should apply values from theme', () => {
    const textTheme = {
      Image: {
        placeholderStyle: {
          backgroundColor: 'red',
        },
      },
    };
    const { queryByTestId } = renderWithTheme(
      <Image source={{ uri: FAKE_URI }} />,
      textTheme
    );
    const placeholder = queryByTestId('RNE__Image__placeholder');
    expect(placeholder.props.style.backgroundColor).toBe('red');
  });

  it('should render without the transition', () => {
    const { queryByTestId } = renderWithTheme(
      <Image source={{ uri: FAKE_URI }} transition={false} />
    );
    const image = queryByTestId('RNE__Image');
    expect(image.props.onLoad).not.toBeNull();
  });

  it('should render PlaceholderContent', () => {
    const { queryByTestId } = renderWithTheme(
      <Image
        source={{ uri: FAKE_URI }}
        PlaceholderContent={<Text testID="text">MyText</Text>}
      />
    );
    const imageText = queryByTestId('text');
    expect(imageText.props.children).toBe('MyText');
  });

  it('should apply value from style prop', () => {
    const { queryByTestId } = renderWithTheme(
      <Image source={{ uri: FAKE_URI }} style={{ tintColor: 'red' }} />
    );
    const image = queryByTestId('RNE__Image');
    expect(image.props.style.tintColor).toBe('red');
  });

  it('should apply value from childrenContainerStyle prop', () => {
    const { queryByTestId } = renderWithTheme(
      <Image
        source={{ uri: FAKE_URI }}
        childrenContainerStyle={{
          borderWidth: 1,
          borderColor: 'red',
        }}
      />
    );
    const childContainer = queryByTestId('RNE__Image__children__container');

    expect(childContainer.props.style).toMatchObject({
      borderWidth: 1,
      borderColor: 'red',
    });
  });

  it('static methods should be present', () => {
    expect(Image.getSize).toBe(RNImage.getSize);
    expect(Image.getSizeWithHeaders).toBe(RNImage.getSizeWithHeaders);
    expect(Image.prefetch).toBe(RNImage.prefetch);
    expect(Image.abortPrefetch).toBe(RNImage.abortPrefetch);
    expect(Image.queryCache).toBe(RNImage.queryCache);
    expect(Image.resolveAssetSource).toBe(RNImage.resolveAssetSource);
  });
});
