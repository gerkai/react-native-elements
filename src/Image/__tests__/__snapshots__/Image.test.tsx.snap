// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Component should render 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <Image
    onLoad={[Function]}
    replaceTheme={[Function]}
    source={
      Object {
        "uri": "https://i.imgur.com/0y8Ftya.jpg",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="customTestId"
    theme={
      Object {
        "colors": Object {
          "black": "#242424",
          "disabled": "hsl(208, 8%, 90%)",
          "divider": "#bcbbc1",
          "error": "#ff190c",
          "grey0": "#393e42",
          "grey1": "#43484d",
          "grey2": "#5e6977",
          "grey3": "#86939e",
          "grey4": "#bdc6cf",
          "grey5": "#e1e8ee",
          "greyOutline": "#bbb",
          "platform": Object {
            "android": Object {
              "error": "#f44336",
              "grey": "rgba(0, 0, 0, 0.54)",
              "primary": "#2196f3",
              "searchBg": "#dcdce1",
              "secondary": "#9C27B0",
              "success": "#4caf50",
              "warning": "#ffeb3b",
            },
            "default": Object {
              "error": "#ff3b30",
              "grey": "#7d7d7d",
              "primary": "#007aff",
              "searchBg": "#dcdce1",
              "secondary": "#5856d6",
              "success": "#4cd964",
              "warning": "#ffcc00",
            },
            "ios": Object {
              "error": "#ff3b30",
              "grey": "#7d7d7d",
              "primary": "#007aff",
              "searchBg": "#dcdce1",
              "secondary": "#5856d6",
              "success": "#4cd964",
              "warning": "#ffcc00",
            },
            "web": Object {
              "error": "#ff190c",
              "grey": "#393e42",
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#ca71eb",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          },
          "primary": "#2089dc",
          "searchBg": "#303337",
          "secondary": "#ca71eb",
          "success": "#52c41a",
          "warning": "#faad14",
          "white": "#ffffff",
        },
      }
    }
    transitionDuration={360}
    updateTheme={[Function]}
  />
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    pointerEvents="none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#bdbdbd",
          "justifyContent": "center",
        }
      }
      testID="RNE__Image__placeholder"
    />
  </View>
  <View
    style={Object {}}
    testID="RNE__Image__children__container"
  />
</View>
`;
